### ISSUE TITLE
[DISCUSSION] Read from a pipe with timeout

### ISSUE THREAD

**Issue by pdonadeo** [c-cube/lwt-pipe#7]
This is the result of my experiments.

**TL;DR;** I WIN! I IMPLEMENTED A SUPER USEFUL FUNCTION! :smile:

Longer and more serious version.

First I ~~stole~~ got inspired by the `values_avalaible` function present in Async. What I need is a way to read from a pipe, blocking if nothing is available like a normal `read`, but not forever. If you think about it it's more or less a "high level" form of Unix `select`.

The code was convoluted and @c-cube asked me a revision, and he was right.

So I re-implemented it in [another branch](https://github.com/pdonadeo/lwt-pipe/tree/data-available-2) using a mutable list of "waiters". This somewhat works but there are at least two problems here:
1. waiters and readers are in different list so... when a thread writes, who is supposed to be signaled? The readers, the waiters or... both? I decided to wake up all the waiters, clear the list and continue with the normal write, waking up a reader
2. this works with one reader/waiter and many writers but when multiple threads are reading/waiting the expected result is... undefined?

Not good at all.

So I decided to give up with this funny function and implement a `read_with_timeout` in the [read-with-timeout](https://github.com/pdonadeo/lwt-pipe/tree/read-with-timeout) branch. The patch is this one: pdonadeo/lwt-pipe@071e12ebc32dbf1b8e906b74a22d4b34286e3102

The idea here is to store in the readers Queue a `bool ref` with Lwt.u. This bool becomes true when the timeout is reached. In the normal `read` it's stored as `false` and never changed. In the "read with timeout" it's stored `false` at the beginning but another thread starts concurrently and sets it to `true`.

The `write`: when a reader is available from the Queue it's popped as usual. If the timeout expired that reader is thrown away and the `write_step` is called again recursively.

This function has a clear semantics, I believe, with no ambiguity and only one queue of readers.

What do you think?

I want to discuss here before pushing a request.

**Comment by c-cube** [comment]
That seems much simpler indeed, and doesn't affect the existing functionality too much. Thank you for your efforts! If you make a PR I think I'll merge (and then harmonize the style a bit :smile:)

If you use lwt-pipe a bit seriously, it'd be very good if you could contribute tests / stress tests, too.

**Comment by pdonadeo** [comment]
Ok, I'm going to submit the PR.

About tests: of course, as soon as I find a project I plan to use lwt-pipe and Lwt more in general.
